### ISSUE TITLE
BUG - sev-2 - [iOS/Android/All] - Increase in session refresh error rates (which forces users to re-authenticate)

### ISSUE THREAD

**Issue by TKDickson** [department-of-veterans-affairs/va-mobile-app#9615]
## What happened?

We received a report from folks in the Identity DSVA channel that there's been an increase in session refresh error rates, since about mid-July.
<img width="1753" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNW5sQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1e1bc5112c48a1007f0107cb75c45fa6a574acdb/image.png" alt="image.png" />
https://vagov.ddog-gov.com/metric/explorer?fromUser=true&start=1717444052000&end=1725392852000&paused=false

The user experience for this is logging into the app, and then (within 45 days, with biometrics enabled) being logged out of the app and having to use UN/PW to log in again. This has been reported by at least one production user.

## Specs:

- Device:
- OS:
- User Account:
- Accessibility State:

## Steps to Reproduce

unknown

## Desired behavior

Should be maintaining login session tokens for the full 45 days, so folks can log in easier with biometrics

## Bug Severity - BE SURE TO ADD THE SEVERITY LABEL
 
See [Bug Tracking](https://department-of-veterans-affairs.github.io/va-mobile-app/docs/QA#issue-severity) for details on severity levels

- Impact: High 
- Frequency: Low 

- 2 - High

## Screen shot(s) and additional information

<details>
  <summary>Full JSON response for services related to issue (expand/collapse)</summary>
  

</details>

## Ticket Checklist
- [ ] Steps to reproduce are defined
- [ ] Desired behavior is added
- [ ] Labels added (front-end, back-end, global, Health, relevant feature, accessibility, etc)
- [ ] Estimate of 1 added (for front-end tickets)
- [ ] Added either to the relevant feature epic (if found during new feature implementation), or the relevant team's bug epic ([Global](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8302), [Health & Benefits](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8303), [API](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8304), [QART](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8305))

**Comment by timwright12** [comment]
Slack thread for additional context: https://dsva.slack.com/archives/C04LFGY2SJK/p1725392845790499

**Comment by Sparowhawk** [comment]
There is a spike of logNonFatalErrorToFirebase(err, `startBiometricsLogin: ${authNonFatalErrorString}`)

So looking through that non fatal I see that the spike started with 2.30 at least compared with the previous versions. ~780 vs previous versions of it being mid 200s and then exploding at 2.32 where it jumps to 23k for Android and iOS it was a slow ramp up starting in 2.24 ~300 and steadily gaining more each release and again exploding in 2.32 to 38k.
I went through all the prs for 2.30, 2.31, 2.32, and nothing changed that would account for that spike.
I did find that the logging for logNonFatalErrorToFirebase(err, `startBiometricsLogin: ${authNonFatalErrorString}`)
occurs if refreshToken = await retrieveRefreshToken() fails
which means this method failed

const retrieveRefreshToken = async (): Promise<string | undefined> => {
  console.debug('retrieveRefreshToken')
  const result = await Promise.all([
    AsyncStorage.getItem(REFRESH_TOKEN_ENCRYPTED_COMPONENT_KEY),
    Keychain.getInternetCredentials(KEYCHAIN_STORAGE_KEY),
  ])
  const reconstructedToken = result[0] && result[1] ? `${result[0]}.${result[1].password}.V0` : undefined

  if (reconstructedToken) {
    await logAnalyticsEvent(Events.vama_login_token_get(true))
  } else {
    await logAnalyticsEvent(Events.vama_login_token_get(false))
  }

  return reconstructedToken
}

**Comment by Sparowhawk** [comment]
Looking further I found there is also a non fatal for saving the refresh token, but it’s numbers didn’t spike so that likely indicates that saving is done correctly.
Therefore I think the issue is contained to the method I posted above, which makes no sense if it was saved correctly as that code hasn’t changed.
So currently waiting on the data dog analytics for this, but the only thing I can think of is maybe the storage size of the nonce is greater than this comment. “Biometric storage has a max storage size of 384 bytes.”

**Comment by Sparowhawk** [comment]
so I guess my next question is do we have a spike for vama_login_token_get with a value of false?
depending on if there is a spike or not means there could be a different problem
one would be with the upstream, the other would be with Keychain or Async Storage

**Comment by Sparowhawk** [comment]
also knowing if there was a spike in vama_login_token_store being false would also be useful

**Comment by Sparowhawk** [comment]
Taking a look at the libraries for Async Storage and Keychain nothing jumps out in their issues that would be causing this problem, but it could still be possible.

**Comment by Sparowhawk** [comment]
I checked the version history for both Keychain and Async Storage for which version we used to see if it was possible we updated to an unstable version. Looks like Keychain and Async Storage both didn't change versions from when we had stable releases with out the spike and immediately after the spike.

This leads me to further believe that it isn't a library problem

We have updated both since then and there are currently new versions available for both that haven't been upgraded yet.

**Comment by Sparowhawk** [comment]
Looking further at the logs I think this might lead to more for Android

Wrapped error: javax.crypto.IllegalBlockSizeException

iOS was not as helpful The user name or passphrase you entered is not correct.

**Comment by Sparowhawk** [comment]
with that error I was able to find this thread  https://github.com/oblador/react-native-keychain/issues/262 

and it looks like a solution is to specify 

 storage: Keychain.STORAGE_TYPE.AES, in the keychain options as biometrics picks one based off the settings and RSA and the other types sometimes throws this error but AES seems to be reliable

**Comment by Sparowhawk** [comment]
@theodur Waiting on Theo's upgrade ticket for the library so we can specify the biometrics AES option instead of letting the device pick

**Comment by rbontrager** [comment]
@Sparowhawk Your fix fixed the original issue but brought on another issue.  It looks like the biometric authentication is occurring during the first time login as well (when it shouldn't). Still just for android. So if you uninstall/re-install the app, login to staging, and press yes to use biometrics a biometric authentication popup is appearing in front of the onboarding screen.

**Comment by theodur** [comment]
> [@theodur](https://github.com/theodur) Waiting on Theo's upgrade ticket for the library so we can specify the biometrics AES option instead of letting the device pick

@Sparowhawk The [PR](https://github.com/department-of-veterans-affairs/va-mobile-app/pull/10061) with the upgrade changes was merged yesterday

**Comment by jennb33** [comment]
1/8/2025 - The MFS team is taking this ticket on for Sprint work.
cc: @ATeal @donmccaughey

**Comment by matt-guest-wilcore** [comment]
Judging from the comments on this ticket it looks like there has been work done on this issue, it would be helpful to see if it's still persisting. I don't have access to DataDog but if I could get confirmation that this issue is still occurring then that would be helpful.

Also, I don't have a physical device to test on but I've been simulating biometrics with the Android emulator and haven't run into any issues yet. Will keep testing.

**Comment by matt-guest-wilcore** [comment]
Heard back from Joe and he confirmed this is still an issue, with a spike in errors occurring November and December of last year. He also mentioned how he had first-hand experience with the issue recently.

Adding chart for reference.

![Image](https://github.com/user-attachments/assets/9b5dec6c-a071-4903-b175-b13ef62eed9d)

**Comment by jennb33** [comment]
2/3/2025 - Moving the ticket to blocked, pending Joe Niquette's return on 2/5/2025 (per @matt-guest-wilcore)

**Comment by matt-guest-wilcore** [comment]
Based on the discussions here in this ticket, it looks like react-native-keychain could be the root cause. I noticed [PR #10040](https://github.com/department-of-veterans-affairs/va-mobile-app/pull/10040) was opened to address this, but was never merged. However, per [this comment](https://github.com/department-of-veterans-affairs/va-mobile-app/issues/9615#issuecomment-2477071869), it seems to be a follow-up to [PR #10061](https://github.com/department-of-veterans-affairs/va-mobile-app/pull/10061), which has already been merged.

I thinking that if #10040 is merged, it’ll likely resolve this issue.

**Comment by jennb33** [comment]
2/6/2025 - Joe Niquette reproduced this error and wants to review with his team, then @matt-guest-wilcore will run the reproduction of the issue with the #10040

**Comment by jason-dehaan** [comment]
2/10/25 Matt to pair with Alex to test Android

**Comment by jason-dehaan** [comment]
2/11/25 Blocked until Matt and Alex can pair

**Comment by jason-dehaan** [comment]
2/12/25 Alex to take a shot at reproducing this. Matt cannot get the error to reproduce.

**Comment by matt-guest-wilcore** [comment]
2/13/25 - We haven’t been able to reproduce #9615 ourselves, but based on the open PR and discussions, we believe the issue is still tied to react-native-keychain. We’re planning to move this to peer review, get it merged, and then monitor analytics to see if it resolves the problem over time.
